runtime/src/local_runqueue_mts.c

```diff
@@ -0,0 +1,281 @@
+ * @param sandbox to delete
+ */
+static void
+local_runqueue_mts_delete(struct sandbox *sandbox)
+{
+	assert(sandbox != NULL);
+	struct perworker_module_sandbox_queue *pwm = &sandbox->module->pwm_sandboxes[worker_thread_idx];
+
+	/* Delete the sandbox from the corresponding Per-Worker-Module queue */
+	if (priority_queue_delete_nolock(pwm->sandboxes, sandbox) == -1)
+		panic("Tried to delete sandbox %lu from PWM queue, but was not present\n", sandbox->id);
+
+	struct priority_queue *destination_queue = local_runqueue_mts_default;
+	/* Because of the race, the following if clause does not work properly, so just trying both queues */
+	/*
+	if (sandbox->module->remaining_budget > 0) {
+	        destination_queue = local_runqueue_mts_guaranteed;
+	}
+	*/
+
+	/* Delete the PWM from the local runqueue completely if pwm is empty, re-add otherwise to heapify */
+	if (priority_queue_delete_nolock(destination_queue, pwm) == -1) {
+		// TODO: Apply the composite way of PQ deletion O(logn)
+		destination_queue = local_runqueue_mts_guaranteed;
+
+		if (priority_queue_delete_nolock(destination_queue, pwm) == -1)
+			panic("Tried to delete a PWM of %s from local runqueue, but was not present\n",
+			      pwm->module->name);
+	}
+	// debuglog("Deleted the PWM from the Worker Local runqueue - %s to Heapify", QUEUE_NAME);
+
+	/* Add the PWM back to the local runqueue if it still has other sandboxes inside */
+	if (priority_queue_length_nolock(pwm->sandboxes) > 0) {
+		priority_queue_enqueue_nolock(destination_queue, pwm);
+		// debuglog("Added the PWM back to the Worker Local runqueue - %s to Heapify", QUEUE_NAME);
+	} else if (pwm->module->replenishment_period > 0) {
+		struct module_timeout *dummy = NULL;
+		priority_queue_dequeue_nolock(worker_thread_timeout_queue, (void **)&dummy);
+		assert(dummy != NULL);
+	}
+}
+
+/**
+ * This function determines the next sandbox to run.
+ * This is the head of the runqueue
+ *
+ * Execute the sandbox at the head of the thread local runqueue
+ * @return the sandbox to execute or NULL if none are available
+ */
+struct sandbox *
+local_runqueue_mts_get_next()
+{
+	/* Get the deadline of the sandbox at the head of the local request queue */
+	struct perworker_module_sandbox_queue *next_pwm = NULL;
+	struct priority_queue *                dq       = local_runqueue_mts_guaranteed;
+
+	int rc = priority_queue_top_nolock(dq, (void **)&next_pwm);
+	while (rc != -ENOENT && next_pwm->module->remaining_budget <= 0) {
+		local_runqueue_mts_demote(next_pwm);
+		debuglog("Demoted '%s' locally", next_pwm->module->name);
+		next_pwm->mt_class = MT_DEFAULT;
+
+		rc = priority_queue_top_nolock(dq, (void **)&next_pwm);
+	}
+
+
+	if (rc == -ENOENT) {
+		dq = local_runqueue_mts_default;
+		rc = priority_queue_top_nolock(dq, (void **)&next_pwm);
+
+		if (rc == -ENOENT) return NULL;
+	}
+
+	struct sandbox *next_sandbox = NULL;
+
+	priority_queue_top_nolock(next_pwm->sandboxes, (void **)&next_sandbox);
+
```

Comment:
I'd imagine you want to remove empty line since the assertion is affine to the function. Up to you.